rust/lance/src/dataset/write/merge_insert/exec/write.rs

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use datafusion::common::Result as DFResult;
+use datafusion::physical_plan::metrics::{BaselineMetrics, ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::{
+    execution::{SendableRecordBatchStream, TaskContext},
+    physical_plan::{
+        execution_plan::{Boundedness, EmissionType},
+        DisplayAs, ExecutionPlan, PlanProperties,
+    },
+};
+use datafusion_physical_expr::{EquivalenceProperties, Partitioning};
+
+use crate::{
+    dataset::write::merge_insert::{exec::MergeInsertMetrics, MergeInsertParams},
+    Dataset,
+};
+
+use super::MERGE_STATS_SCHEMA;
+
+/// Inserts new rows and updates existing rows in the target table.
+///
+/// This does the actual write.
+///
+/// This is implemented by moving updated rows to new fragments. This mode
+/// is most optimal when updating the full schema.
+///
+/// It returns a single batch, containing the statistics.
+#[derive(Debug)]
+pub struct FullSchemaMergeInsertExec {
+    input: Arc<dyn ExecutionPlan>,
+    dataset: Arc<Dataset>,
+    params: MergeInsertParams,
+    properties: PlanProperties,
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl FullSchemaMergeInsertExec {
+    pub fn try_new(
+        input: Arc<dyn ExecutionPlan>,
+        dataset: Arc<Dataset>,
+        params: MergeInsertParams,
+    ) -> DFResult<Self> {
+        let properties = PlanProperties::new(
+            EquivalenceProperties::new((*MERGE_STATS_SCHEMA).clone()),
+            Partitioning::UnknownPartitioning(1),
+            EmissionType::Final,
+            Boundedness::Bounded,
+        );
+
+        Ok(Self {
+            input,
+            dataset,
+            params,
+            properties,
+            metrics: ExecutionPlanMetricsSet::new(),
+        })
+    }
+}
+
+impl DisplayAs for FullSchemaMergeInsertExec {
+    fn fmt_as(
+        &self,
+        t: datafusion::physical_plan::DisplayFormatType,
+        f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        write!(f, "FullSchemaMergeInsertExec")
+    }
+}
+
+impl ExecutionPlan for FullSchemaMergeInsertExec {
+    fn name(&self) -> &str {
+        "FullSchemaMergeInsertExec"
+    }
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> arrow_schema::SchemaRef {
+        (*MERGE_STATS_SCHEMA).clone()
+    }
```

Comment:
I think that instead of using the data output for this, it would be easier to just provide accessor methods on the `FullSchemaMergeInsertExec`. We can downcast the execution plan after running it and use those accessor methods to extract the results.